skribenten-web/frontend/src/Brevredigering/LetterEditor/actions/toggleBulletList.ts

```diff
@@ -0,0 +1,436 @@
+import type { Draft } from "immer";
+import { produce } from "immer";
+import { isEqual } from "lodash";
+
+import type { Content, ItemList, ParagraphBlock, TextContent } from "~/types/brevbakerTypes";
+
+import type { Action } from "../lib/actions";
+import type { LetterEditorState } from "../model/state";
+import { isTextContent } from "../model/utils";
+import { newItemList, newParagraph } from "./common";
+import type { ItemContentIndex, LiteralIndex } from "./model";
+
+export const toggleBulletList: Action<LetterEditorState, [literalIndex: LiteralIndex]> = produce(
+  (draft, literalIndex) => {
+    const block = draft.redigertBrev.blocks[literalIndex.blockIndex];
+    if (block.type !== "PARAGRAPH") {
+      return;
+    }
+
+    draft.isDirty = true;
+    const theContentTheUserIsOn = block.content[literalIndex.contentIndex];
+    if (isTextContent(theContentTheUserIsOn)) {
+      toggleBulletListOn(draft, literalIndex);
+    } else if (theContentTheUserIsOn.type === "ITEM_LIST") {
+      toggleBulletListOff(draft, literalIndex as ItemContentIndex);
+    }
+  },
+);
+
+/**
+ * Når vi lager et punkt, så må vi ta høyde for at det kan være en punktliste før, etter, eller ingen punktliste.
+ * Det kan finnes punktliste i blokk før, blokk etter, eller begge. Det kan også være punktliste i samme blokk, før / etter / begge, på literalen vi er på
+ *
+ * Det vil si at når vi converterer en literal til en ny punktliste, merger vi med nabo-punktlistene, hvis de eksisterer
+ * Så må vi sjekke om det finnes nabo-punktlister på blokk-nivå, og merge de også
+ *
+ * Fordi vi gjør en såpass stor endring i dokument strukturen, Så må vi oppdatere fokuset til editorstaten til å være på rett plass
+ */
+const toggleBulletListOn = (draft: Draft<LetterEditorState>, literalIndex: LiteralIndex) => {
+  //-------------- Convertering av literal til punktliste ----------------
+  const thisBlock = draft.redigertBrev.blocks[literalIndex.blockIndex];
+  const theIdexOfTheContent = literalIndex.contentIndex;
+  const sentence = getSurroundingLiteralsAndVariables(thisBlock.content, theIdexOfTheContent);
+  const sentenceElements = sentence.map((r) => r.element);
+
+  const replacedWithItemList = replaceElementsBetweenIncluding(
+    thisBlock.content,
+    sentence[0].originalIndex,
+    sentence.at(-1)!.originalIndex,
+    newItemList({
+      items: [{ id: null, content: sentenceElements }],
+    }),
+  );
+
+  const { newcontent, newDeleted } = mergeItemLists(replacedWithItemList);
+
+  const newThisParagraph: ParagraphBlock = newParagraph({
```

Comment:
Jeg tror jeg ville gjort inplace modifisering av `thisBlock`, slik at vi ikke får problemer om vi skulle introdusere flere felter på paragraph på et tidspunkt. Uavhengig av det så er det et problem nå med at deletedContent vil miste evt. tidligere slettet content, har pushet test for det.